server/src/main/java/org/opensearch/index/translog/transfer/BaseTranslogTransferManager.java

```diff
@@ -99,46 +101,78 @@ public ShardId getShardId() {
+        WritePriority writePriority
+    ) throws Exception;
+
+    /**
+     * @return true if fileName based tracker needs to be updated, false otherwise
+     */
+    abstract boolean updateFileNameTransferTracker();
+
     public boolean transferSnapshot(TransferSnapshot transferSnapshot, TranslogTransferListener translogTransferListener)
         throws IOException {
-        List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
-        Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
         long metadataBytesToUpload;
         long metadataUploadStartTime;
         long uploadStartTime;
         long prevUploadBytesSucceeded = remoteTranslogTransferTracker.getUploadBytesSucceeded();
         long prevUploadTimeInMillis = remoteTranslogTransferTracker.getTotalUploadTimeInMillis();
+        int totalFilesCount = transferSnapshot.getTranslogTransferMetadata().getCount();
+        List<Exception> exceptionList = new ArrayList<>(totalFilesCount);
+        Set<TranslogCheckpointSnapshot> generationalSnapshotList = new HashSet<>(totalFilesCount);
 
         try {
-            toUpload.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogFileSnapshots()));
-            toUpload.addAll(fileTransferTracker.exclusionFilter((transferSnapshot.getCheckpointFileSnapshots())));
-            if (toUpload.isEmpty()) {
+            generationalSnapshotList.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogCheckpointSnapshots()));
+            if (generationalSnapshotList.isEmpty()) {
                 logger.trace("Nothing to upload for transfer");
                 return true;
             }
 
-            fileTransferTracker.recordBytesForFiles(toUpload);
+            fileTransferTracker.recordBytesForFiles(generationalSnapshotList);
             captureStatsBeforeUpload();
-            final CountDownLatch latch = new CountDownLatch(toUpload.size());
-            LatchedActionListener<TransferFileSnapshot> latchedActionListener = new LatchedActionListener<>(
+            final CountDownLatch latch = new CountDownLatch(generationalSnapshotList.size());
+            LatchedActionListener<TranslogCheckpointSnapshot> latchedActionListener = new LatchedActionListener<>(
                 ActionListener.wrap(fileTransferTracker::onSuccess, ex -> {
-                    assert ex instanceof FileTransferException;
+                    assert ex instanceof TranslogTransferException;
                     logger.error(
                         () -> new ParameterizedMessage(
-                            "Exception during transfer for file {}",
-                            ((FileTransferException) ex).getFileSnapshot().getName()
+                            "Exception during transfer for translog generation {}",
+                            ((TranslogTransferException) ex).getFileSnapshot().getGeneration()
```

Comment:
Lets enhance the log to print which file exactly failed tlog or generation since the case without metadata will continue to upload 2 files.